java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,239 @@
+        handleSpec: Plan.HandleConnection,
+        holder: HandleHolder
+    ) = when (handleSpec.type) {
+        is SingletonType<*> ->
+            entityHandleManager().createSingletonHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        is CollectionType<*> ->
+            entityHandleManager().createSetHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        else -> throw Exception("Unknown type ${handleSpec.type}")
     }
 
     override suspend fun stopArc(partition: Plan.Partition) {
-        // TODO: not implemented yet
+        val arcId = partition.arcId
+        val context = lookupOrCreateArcHostContext(partition)
+        runningArcs[partition.arcId]?.let { context ->
+            when (context.arcState) {
+                ArcState.Running -> stopArcInternal(arcId, context)
+                ArcState.NeverStarted -> stopArcError(context, "Arc $arcId was never started")
+                ArcState.Stopped -> stopArcError(context, "Arc $arcId already stopped")
+                ArcState.Deleted -> stopArcError(context, "Arc $arcId is deleted.")
+            }
+        }
+    }
+
+    suspend fun stopArcError(context: ArcHostContext, message: String) {
+        // TODO: decide how to propagate this
+    }
+
+    suspend fun stopArcInternal(arcId: String, context: ArcHostContext) {
+        for (particleContext in context.particles.values) {
+            stopParticle(particleContext)
+        }
+        maybeCancelResurrection(context)
+        context.arcState = ArcState.Stopped
+        updateArcHostContext(arcId, context)
+    }
+
+    suspend fun stopParticle(context: ParticleContext) {
+        context.particle.onShutdown()
+        context.particleState = ParticleState.Stopped
+        cleanupHandles(context.particle.handles)
+    }
+
+    suspend fun cleanupHandles(handles: HandleHolder) {
+        for ((name, handle) in handles.handles) {
+            // TODO: disconnect/unregister handle
+        }
+        (handles.handles as MutableMap).clear()
     }
 
     override suspend fun isHostForParticle(particle: Plan.Particle) =
         registeredParticles().contains(ParticleIdentifier.from(particle.location))
-}
 
-/**
- * Convert an array of [Particle] class literals to a [MutableList] of [ParticleIdentifier].
- */
-fun Array<out KClass<out Particle>>.toIdentifierList(): MutableList<ParticleIdentifier> =
-    this.map { it.toParticleIdentifier() }.toMutableList()
+    // TODO: hack workaround, replace with portable core version?
+    class TimeImpl : Time() {
+        override val currentTimeNanos: Long
+            get() = System.nanoTime()
+        override val currentTimeMillis: Long
+            get() = System.currentTimeMillis()
+    }
+
+    open fun platformTime(): Time = TimeImpl()
```

Comment:
KDoc would be hugely useful for this. Additionally - this could/should be a `val`.  As an extender of AbstractArcHost, I would wonder why I am allowed to override platformTime, and what it'll be used for.

Reply:
It's needed by the HandleManager now for TTLs I think, ergo, shared code needs someway to get the current time that is platform independent.  Since there's no multiplatform Kotlin equivalent of time APIs right now, it's encapsulated like this.